Simplified/Catalog/NYPLCatalogUngroupedFeed.m

```diff
@@ -27,31 +27,55 @@ @interface NYPLCatalogUngroupedFeed ()
 // attempt to fetch more books will be made.
 static NSUInteger const preloadThreshold = 100;
 
+// The current number of attempts of books fetching until we receive supported books,
+// and the maximum attempts allowed.
+static NSInteger fetchSupportedBooksRetryCount = 0;
+static NSInteger const fetchSupportedBooksRetryThreshold = 15;
+
 @implementation NYPLCatalogUngroupedFeed
 
-+ (void)withURL:(NSURL *)URL
-handler:(void (^)(NYPLCatalogUngroupedFeed *category))handler
-{
++ (void)fetchSupportedBooksWithURL:(NSURL *)URL
+                          provider:(Class <NYPLOPDSFeedProvider>)provider
+                           handler:(void (^)(NYPLCatalogUngroupedFeed *feed))handler {
   if(!handler) {
     @throw NSInvalidArgumentException;
   }
   
-  [NYPLOPDSFeed
+  if (fetchSupportedBooksRetryCount >= fetchSupportedBooksRetryThreshold) {
+    NYPLLOG(@"Retry threshold reached while fetching ungrouped feed.");
+    fetchSupportedBooksRetryCount = 0;
+    handler(nil);
+    return;
+  }
+  
+  fetchSupportedBooksRetryCount += 1;
+  
+  [provider
    withURL:URL
    shouldResetCache:NO
    completionHandler:^(NYPLOPDSFeed *const ungroupedFeed, __unused NSDictionary *error) {
-     if(!ungroupedFeed) {
-       handler(nil);
-       return;
-     }
-     
-     if(ungroupedFeed.type != NYPLOPDSFeedTypeAcquisitionUngrouped) {
-       NYPLLOG(@"Ignoring feed of invalid type.");
-       handler(nil);
-       return;
-     }
+    if(!ungroupedFeed) {
+      fetchSupportedBooksRetryCount = 0;
```

Comment:
from this point of view, and unless the `provider withURL:...` method guarantees to call the completionHandler on the same thread it itself was being called on, this assignment can happen on a different thread than what happens at line 51, causing inconsistencies.

Reply:
I will make this counter and the function `fetchSupportedBooksWithURL` non-static and it should resolve this issue.